page/script/page.js

```diff
@@ -0,0 +1,130 @@
+define(function(require, exports, module) {
+  /*********************
+   * Prevent addition of page/page-* documents with duplicate urls of existing documents
+   * ******************/
+
+
+  var $ = require("jquery");
+  var duplicateUrlString = "Oh-Oh, it looks like another page is already using that URL."
+  var validUrlString = "The URL is not being used, you are good to go."
+  var errorColor = "#a94442"
+  var validColor = "rgb(39, 174, 96)"
+  var saveButtonSelector = ".buttonbar .btn.save, .btn-toolbar .btn.save, .btn.btn-success.wizard-next"
+  var urlFieldSelector = ".alpaca-field [data-alpaca-field-name=urlList]"
+  var urlTextSelector = ".alpaca-field [name=urlList_0_url]"
+  var helpBlockSelector = ".alpaca-helper.help-block"
+  var validUrlClass = "valid-url-message"
+  var validUrlSelector = "." + validUrlClass
+  var duplicateUrlClass = "duplicate-url-message"
+  var duplicateUrlSelector = "." + duplicateUrlClass
+  var branch = Ratchet.observable('branch').get()
+  var latestPagesRequestTime
+  var timer = undefined
+  var docId = null
+
+  function disableSave() {
+    $(saveButtonSelector).attr("disabled", "disabled")
+  }
+  
+  function enableSave() {
+    $(saveButtonSelector).removeAttr("disabled")
+  }
+
+  function clearMessages() {
+    $(duplicateUrlSelector + ", " + validUrlSelector).remove()
+  }
+
+  function setUrlInvalid() {
+    disableSave()
+    //preemptively remove to prevent occasional double insertion
+    clearMessages()
+    //add message text
+    $(urlFieldSelector).find(helpBlockSelector).after(
+      "<p class='" + duplicateUrlClass + "' style='color: " + errorColor + ";'>" + duplicateUrlString + "</p>")
+  }
+
+  function setUrlValid() {
+    enableSave()
+    //remove message text
+    clearMessages()
+    //add message text
+    $(urlFieldSelector).find(helpBlockSelector).after(
+      "<p class='" + validUrlClass + "' style='color: " + validColor + ";'>" + validUrlString + "</p>")
+  }
+
+  function genericErrorLoggerHalter(err) {
+    console.error(err);
+    return false;
+  }
+
+  function queryForPages(url) {
+    var chain = Chain(branch).trap(genericErrorLoggerHalter)
+    return chain.queryNodes({
+      _type: {
+        $regex: "cricket:page(-.*)?"
+      },
+      _doc: {
+        $ne: docId
+       
+      },
+      urlList: {
+        $elemMatch: {
+          url: url
+        }
+      },
+      active: 'y'
+    })
+  }
+
+  function findIdenticalUrlPages(event) {
+    latestPagesRequestTime = new Date().getTime()
+    var pagesRequestTime = new Date().getTime()
+    var url
+
+    //disable early as response timing varies
+    disableSave()
+    clearMessages()
+
+    //cancel previous request if new event comes in before it completes
+    if (timer) {
+      clearTimeout(timer)
+    }
+
+    //schedule scrape of url value after current thread has finished
+    //(cut/paste event-driven changes do not reflect until then)
```

Comment:
Nice, didnt even think about cut/paste

Reply:
It's slightly irritating (but I decided to just let it be right now since nothing is perfect and they can just modify the JSON manually and skirt the entire validation...) that there is no such thing as an undo or redo event to bind to. A user who pastes into a field triggers the call to `queryNodes`, but if the user then right clicks and selects undo from the context menu, the field can be empty but the page still shows the "invalid" state.

The only way to handle this edge case is having a timeout/interval running endlessly, and check for empty field/clear messaging when it runs if so. This wastes cycles and could result in leakage if we're not diligent cleaning up after ourselves, and maintaining such diligence in future tweaks due to changes in DOM structure or other things we hook into within Cloud CMS. We could add such a thing but I decided it was better to leave out for now and deal with it later if it's really an issue.

I also am not fond of the fact that the Save button is disabled temporarily even just depressing a shift key or similar, but didn't feel like it was worth handling each and every special/modifier key across platforms or loading in more library overhead just for a tiny UX win. Again, if we need to add it, we can add it later on. Trying to enhance things progressively, not overkill. But adding one more event to bind to is a nice little win for one additional word.